vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs

```diff
@@ -22,6 +22,34 @@ public static IEdmModel BuildEdmModel(Type ApiContextType)
                 builder.AddEntitySet(property.Name, entity);
             }
 
+            // Get the actions and functions into the model
+            var publicMethods = ApiContextType.GetMethods(BindingFlags.Public | BindingFlags.Instance);
+            foreach (var method in publicMethods)
+            {
+                if (!method.IsSpecialName)
+                {
+                    var entityClrType = TypeHelper.GetImplementedIEnumerableType(method.ReturnType) ?? method.ReturnType;
+                    ProcedureConfiguration configuration;
+
+                    if (entityClrType.Name != "Void")
+                    {
+                        var returnType = builder.AddEntityType(entityClrType);
+                        configuration = builder.Function(method.Name);
+                        configuration.ReturnType = returnType;
+                    }
+                    else
+                    {
+                        configuration = builder.Action(method.Name);
```

Comment:
Action can also have a return type, we shall define whether an operation is side effecting to differentiate between action and function.